crates/ty_python_semantic/resources/mdtest/import/nonstandard_conventions.md

```diff
@@ -0,0 +1,607 @@
+
+> If an `__init__.pyi` for `mypackage` contains a `from...import` targetting a submodule, then that
+> submodule should be available as an attribute of `mypackage`.
+
+## Relative `from` Import of Direct Submodule in `__init__`
+
+The `from . import submodule` idiom in an `__init__.pyi` is fairly explicit and we should definitely
+support it.
+
+`mypackage/__init__.pyi`:
+
+```pyi
+from . import imported
+```
+
+`mypackage/imported.pyi`:
+
+```pyi
+X: int = 42
+```
+
+`mypackage/fails.pyi`:
+
+```pyi
+Y: int = 47
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+reveal_type(mypackage.imported.X)  # revealed: int
+# error: "has no attribute `fails`"
+reveal_type(mypackage.fails.Y)  # revealed: Unknown
+```
+
+## Relative `from` Import of Direct Submodule in `__init__` (Non-Stub Check)
+
+`mypackage/__init__.py`:
+
+```py
+from . import imported
+```
+
+`mypackage/imported.py`:
+
+```py
+X: int = 42
+```
+
+`mypackage/fails.py`:
+
+```py
+Y: int = 47
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+reveal_type(mypackage.imported.X)  # revealed: int
+# error: "has no attribute `fails`"
+reveal_type(mypackage.fails.Y)  # revealed: Unknown
+```
+
+## Absolute `from` Import of Direct Submodule in `__init__`
+
+If an absolute `from...import` happens to import a submodule, it works just as well as a relative
+one.
+
+`mypackage/__init__.pyi`:
+
+```pyi
+from mypackage import imported
+```
+
+`mypackage/imported.pyi`:
+
+```pyi
+X: int = 42
+```
+
+`mypackage/fails.pyi`:
+
+```pyi
+Y: int = 47
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+reveal_type(mypackage.imported.X)  # revealed: int
+# error: "has no attribute `fails`"
+reveal_type(mypackage.fails.Y)  # revealed: Unknown
+```
+
+## Absolute `from` Import of Direct Submodule in `__init__` (Non-Stub Check)
+
+`mypackage/__init__.py`:
+
+```py
+from mypackage import imported
+```
+
+`mypackage/imported.py`:
+
+```py
+X: int = 42
+```
+
+`mypackage/fails.py`:
+
+```py
+Y: int = 47
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+reveal_type(mypackage.imported.X)  # revealed: int
+# error: "has no attribute `fails`"
+reveal_type(mypackage.fails.Y)  # revealed: Unknown
+```
+
+## Import of Direct Submodule in `__init__`
+
+An `import` that happens to import a submodule does not expose the submodule as an attribute. (This
+is an arbitrary decision and can be changed easily!)
+
+`mypackage/__init__.pyi`:
+
+```pyi
+import mypackage.imported
+```
+
+`mypackage/imported.pyi`:
+
+```pyi
+X: int = 42
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+# error: "has no attribute `imported`"
+reveal_type(mypackage.imported.X)  # revealed: Unknown
+```
+
+## Import of Direct Submodule in `__init__` (Non-Stub Check)
+
+`mypackage/__init__.py`:
+
+```py
+import mypackage.imported
+```
+
+`mypackage/imported.py`:
+
+```py
+X: int = 42
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+# error: "has no attribute `imported`"
+reveal_type(mypackage.imported.X)  # revealed: Unknown
```

Comment:
I still think it would be great to add a TODO comment here (https://github.com/astral-sh/ruff/pull/20855#discussion_r2429085956)

Since this is what the original examples in github.com/astral-sh/ty/issues/133 were about, I think this PR arguably doesn't actually close that issue -- rather than being about implicit imports of submodules in `__init__.py` files (what that issue was originally about), to my mind this PR is really about carving out exemptions to the explicit re-export convention for stub files in `__init__.pyi` files. It's still a good and important change for us to make! But something slightly different to what github.com/astral-sh/ty/issues/133 was originally about.